atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/LocalWriteBufferTest.java

```diff
@@ -0,0 +1,221 @@
+
+        assertThat(buffer.getLocalWritesForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, VALUE_1, CELL_2, VALUE_2));
+        assertThat(buffer.getChangeMetadataForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, METADATA_1));
+    }
+
+    @Test
+    public void canOverwriteValueAndMetadata() {
+        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(CELL_1, VALUE_1), ImmutableMap.of(CELL_1, METADATA_1));
+        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(CELL_1, VALUE_2), ImmutableMap.of(CELL_1, METADATA_2));
+
+        assertThat(buffer.getLocalWritesForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, VALUE_2));
+        assertThat(buffer.getChangeMetadataForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, METADATA_2));
+    }
+
+    @Test
+    public void writingWithoutMetadataRemovesMetadata() {
+        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(CELL_1, VALUE_1), ImmutableMap.of(CELL_1, METADATA_1));
+        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(CELL_1, VALUE_2), ImmutableMap.of());
+
+        assertThat(buffer.getLocalWritesForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, VALUE_2));
+        assertThat(buffer.getChangeMetadataForTable(TABLE)).isEmpty();
+    }
+
+    @Test
+    public void canStoreValuesAndMetadataForMultipleCellsAndTables() {
+        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(CELL_1, VALUE_1), ImmutableMap.of(CELL_1, METADATA_1));
+        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(CELL_2, VALUE_2), ImmutableMap.of(CELL_2, METADATA_2));
+        buffer.putLocalWritesAndMetadata(TABLE2, ImmutableMap.of(CELL_1, VALUE_1), ImmutableMap.of(CELL_1, METADATA_1));
+        buffer.putLocalWritesAndMetadata(TABLE2, ImmutableMap.of(CELL_2, VALUE_2), ImmutableMap.of(CELL_2, METADATA_2));
+
+        assertThat(buffer.getLocalWritesForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, VALUE_1, CELL_2, VALUE_2));
+        assertThat(buffer.getLocalWritesForTable(TABLE2))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, VALUE_1, CELL_2, VALUE_2));
+        assertThat(buffer.getChangeMetadataForTable(TABLE))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, METADATA_1, CELL_2, METADATA_2));
+        assertThat(buffer.getChangeMetadataForTable(TABLE2))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, METADATA_1, CELL_2, METADATA_2));
+    }
+
+    @Test
+    public void cannotPutMetadataWithoutAssociatedWrite() {
+        assertThatLoggableExceptionThrownBy(() ->
+                        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(), ImmutableMap.of(CELL_1, METADATA_1)))
```

Comment:
We should add one additional test here, where the `values` map has differing entries (CELL_2, VALUE_1). Both are important to test, as the former would catch breaks in `hashcode`/`equals`